File: 112.MeetintheMiddleAlgo.py
# The Meet in the Middle algorithm is a divide-and-conquer 
# strategy used to solve problems by splitting the input into two halves, solving each half independently, and then combining the results efficiently.
# It’s especially useful when brute-force takes too long — 
# typically reducing O(2ⁿ) complexity to O(2ⁿ/²).

# Explanation

# When dealing with problems like subset sums or combinations 
# where we must try all possibilities, directly generating all 
# subsets of n elements would take 2^n time.
# Instead, we:

# Divide the array into two halves.

# Compute all subset sums of both halves.

# Sort one half and use binary search / two pointers to find valid combinations 
# that meet the required condition (like target sum).

# Thus, the complexity improves from O(2ⁿ) → O(2 × 2ⁿ/² × log(2ⁿ/²)), which is manageable for n ≈ 40.

# Example Problem: Subset Sum using Meet in the Middle

# Problem Statement:
# Given an array of integers and a target sum, find if 
# there exists a subset whose sum equals the target.

from bisect import bisect_left

def subset_sums(arr):
    res = [0]
    for x in arr:
        new_sums = [x + y for y in res]
        res += new_sums
    return res

def meet_in_the_middle(nums, target):
    n = len(nums)
    left = nums[:n//2:]
    right = nums[n//2:]

    left_sums = subset_sums(left)
    right_sums = subset_sums(right)
    right_sums.sort()

    for s in left_sums:
        needed = target - s
        idx = bisect_left(right_sums, needed)
        if idx < len(right_sums) and right_sums[idx] == needed:
            return True
    return False

nums = [3, 34, 4, 12, 5, 2]
target = 9

result = meet_in_the_middle(nums, target)

print("Subset sum equal to target exists?", result)

File: 112.MeetintheMiddleAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio        */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

public class MeetInTheMiddle {

    // Function to generate all possible subset sums of a given list
    static List<Long> subsetSums(long[] arr) {
        List<Long> res = new ArrayList<>();
        res.add(0L); // Start with empty subset sum = 0

        // For every number, add it to all existing subset sums
        for (long x : arr) {
            List<Long> newSums = new ArrayList<>();
            for (long y : res) newSums.add(x + y); // Add current element to all previous sums
            res.addAll(newSums); // Combine with previous results
        }
        return res;
    }

    // Function implementing the Meet in the Middle algorithm
    static boolean meetInMiddle(long[] nums, long target) {
        int n = nums.length;
        long[] left = Arrays.copyOfRange(nums, 0, n / 2); // Split into left half
        long[] right = Arrays.copyOfRange(nums, n / 2, n); // Split into right half

        List<Long> leftSums = subsetSums(left);   // All subset sums of left half
        List<Long> rightSums = subsetSums(right); // All subset sums of right half
        Collections.sort(rightSums);              // Sort right half for binary search

        // Check if any pair of (sum_left, sum_right) meets the target
        for (long s : leftSums) {
            long needed = target - s;
            int idx = Collections.binarySearch(rightSums, needed); // Binary search
            if (idx >= 0) return true; // Found a valid subset pair
        }
        return false; // No subset pair found
    }

    public static void main(String[] args) throws IOException {
        // Example input
        long[] nums = {3, 34, 4, 12, 5, 2};
        long target = 9;

        // Running the function
        boolean result = meetInMiddle(nums, target);

        // Printing output
        System.out.println("Subset sum equal to target exists? " + result);
    }
}

/*
Explanation (Line-by-Line):

1. subsetSums() builds all possible sums of subsets using iterative addition.
2. Split nums into two halves for independent processing.
3. Compute leftSums and rightSums — each having 2^(n/2) sums.
4. Sort rightSums to enable fast binary search.
5. For every sum in the left half, compute needed = target - s.
6. Use Collections.binarySearch() to check if needed exists in right half.
7. If found, a valid subset combination exists → return true.
8. Otherwise, return false.

Output:
Subset sum equal to target exists? true

Output Explanation:
Subsets {4, 5} or {3, 4, 2} both sum up to 9.
The algorithm efficiently checks 2^(n/2) combinations per half instead of 2^n.
Hence, the program returns true.

Summary:
Approach: Divide & Conquer (Meet in the Middle)
Use Case: Subset sum, knapsack variants, and counting subsets.
Time Complexity: O(2^(n/2) * log(2^(n/2)))
Space Complexity: O(2^(n/2))
Key Idea: Split → Compute Subsets → Merge with Binary Search
*/
